### ISSUE TITLE
PHP notices when installing modules that have sub-modules

### ISSUE THREAD

**Issue by klonos** [backdrop/backdrop-issues#4640]
Just wanted to quickly raise this (found while testing another PR on my local):

![image](https://user-images.githubusercontent.com/2423362/94183478-28776f00-feab-11ea-9d99-2c6e00252813.png)

The error:

```
Notice: Undefined index: hidden in installer_browser_installation_enable_form()
(line 653 of installer.pages.inc).
```

Steps to reproduce:

- Fire up a fresh installation of Backdrop 1.17.0
- Head to `/admin/config/development/logging` and enable all errors.
- Start installing any module that has sub-modules via the Project Browser/Installer (Metatag and Devel are a couple of example such modules).
- Continue -> errors shown in the "Enable modules" step 👎 
- Try installing any module without any submodules (Entity Plus for example)
- Continue -> no errors shown in the "Enable modules" step 👍

**Comment by indigoxela** [comment]
Confirmed. This seems to be a side effect of https://github.com/backdrop/backdrop-issues/issues/2200

Tested with:

- devel (has submodules)
- backup_migrate (no submodules)
- menu_link_weight (has a hidden submodule)

A [PR](https://github.com/backdrop/backdrop/pull/3330) is available for testing and review

A little `isset()` does the trick.

**Comment by indigoxela** [comment]
Ooops, I just realized, that @klonos had already self-assigned that issue, I totally missed that. :hand_over_mouth:

**Comment by klonos** [comment]
>Ooops, I just realized, that @klonos had already self-assigned that issue, I totally missed that. 🤭

Don't let that even bother you when it comes to me @indigoxela 😄

**Comment by klonos** [comment]
Simple fix indeed ...things work as expected now, w/o any errors 👍 ...RTBC.

**Comment by BWPanda** [comment]
Added a comment to the PR asking about possibly making this a bit more readable...

**Comment by klonos** [comment]
@BWPanda you're right. Please see my comment with suggestion in the PR.

**Comment by indigoxela** [comment]
I'll try to improve code readability.

**Comment by indigoxela** [comment]
Ready for another round of testing and review.

One question popped up that I can't reliably answer: are there situation when `$module->info['project']` is not set?

**Comment by BWPanda** [comment]
Code looks good to me!

**Comment by herbdool** [comment]
@indigoxela if a module is released on b.org then it gets a project line. Otherwise it's possible for it not to be there.

**Comment by klonos** [comment]
Code does look good, but I just tested in the PR sandbox, and got an error:

`Notice: Undefined index: type in installer_browser_installation_enable_form() (line 660 of /var/lib/tugboat/core/modules/installer/installer.pages.inc).`

![image](https://user-images.githubusercontent.com/2423362/98455019-fa58a080-2173-11eb-97d3-3cb535ca5ec7.png)

**Comment by klonos** [comment]
...the culprit in this case was the `metatag_devel` module, which does not have the hidden flag set in its .info file, and at the same time does not define a type (see https://github.com/backdrop-contrib/metatag/blob/1.x-1.x/metatag_devel/metatag_devel.info) ...we currently do not have a way to be enforcing type in .info files of all submodules of contrib projects, but at the same time we should not allow core to choke if that happens.

**Comment by indigoxela** [comment]
> ...the culprit in this case was the metatag_devel module

Yeah, the Metatag module is pretty... ehm...

> we currently do not have a way to be enforcing type in .info files of all submodules of contrib projects, but at the same time we should not allow core to choke if that happens.

Is there consensus that core should catch that? I don't assume that, because it's clearly explained in the [documentation](https://api.backdropcms.org/modules) that the `type` is required:

```
type (Required)
type = module

The type of project. For a module, this will always be "module". Other available types are "theme" or "layout". Although this property is not required to enable the module, it is required to properly package the module on BackdropCMS.org and thus should always be included.
```
Should we really catch obvious violations in contrib modules?

**Comment by klonos** [comment]
> Should we really catch obvious violations in contrib modules?

Perhaps not, but I think that we should do one of the following:

- throw a meaningful "No project type specified for %projectname project in its .info file" error instead of the cryptic/develop-y php notice "undefined index" and add a watchdog notice.
- remove any such offending projects from the `$projects` array in https://github.com/backdrop/backdrop/blob/1.x/core/modules/installer/installer.pages.inc#L645 and add a watchdog notice.

**Comment by indigoxela** [comment]
I'll give it a try, but I have to do some recherche, which additional changes are needed.

Note: the documentation will need an update then:

> Although this property is not required to enable the module...

At least if we decide to prevent enabling the module, if the type property is missing.

Another question: can we really do this in a _bugfix release_? A change record might be necessary.
If it's out of scope of a bugfix release, I'd suggest to do it in a follow-up issue.

**Comment by indigoxela** [comment]
So I checked the current situation...

The violating module is not available in the list of (sub-)modules to enable - as the key is missing. So there's no change in behavior compared to what we do now. My concern that we can't do it in a bugfix release wasn't confirmed.

It makes sense to skip over the violating modules, but log at least its name.

**Comment by indigoxela** [comment]
@BWPanda and @klonos you both started reviewing. Mind to complete it? :smiley: Your requested changes have been done.

**Comment by klonos** [comment]
@indigoxela I've tested this and it works as expected, with the warning logged in watchdog 👍 

I've left a comment in the PR re the use of the word "project" vs "module" + a suggestion to link to https://api.backdropcms.org/modules#info-files so that module builders know what we're talking about.

I don't think that we'd need a change record, but we could tweak our documentation a bit, to explain that modules missing "type" in their .info file will be skipped.

**Comment by Wylbur** [comment]
Sorry to comment on a closed issue, but what to do when I run into this?  I get these errors with the rabbit_hole module. 
https://backdropcms.org/project/rabbit_hole

Will the fixes in core resolve this, or is action on the Rabbit Hole module required.  

Thanks!

**Comment by BWPanda** [comment]
@Wylbur This issue isn't closed, we're still working on it. I looked through the Rabbit Hole module and sub-modules, and I couldn't find any instances where it was missing the `type` value in its info files... Exactly what error are you getting?

@indigoxela @klonos I left a question in the PR about whether the skipping/logging code should be in this function or elsewhere in core. Let me know what you think.

**Comment by Wylbur** [comment]
Here's the error I got on the install page when it listed the newly installed Rabbit Hole module:
```Notice: Undefined index: hidden in installer_browser_installation_enable_form() (line 653 of /home/lenaswan/lsbd/docroot/core/modules/installer/installer.pages.inc).```

**Comment by BWPanda** [comment]
@Wylbur Ah ok. Yes, this module should fix that and I don't think Rabbitt Hole needs to do anything.

**Comment by BWPanda** [comment]
@indigoxela Let's move our discussion here (from the PR)...

The current situation (before this PR) is that modules without a `type` cause errors to be displayed on the site. The PR is doing two things to address that:
1. Skips over modules without a `type`
2. Logs a warning saying that the module is missing `type`

What I'm saying is that *if* we're going to go down the road of logging warnings when we encounter modules without a `type`, let's do it in a more general place (like `_system_rebuild_module_data()`) rather than this function specific to the project installer). The purpose of this is so that more modules get fixed more frequently.

If, however, we're worried that this will cross the line into "enforcing `type` for modules" (which we don't currently do), then I suggest we edit this PR to just skip over these modules and not do any logging at all.

In other words, let's either log these warning all the time, or not at all.

I hope that explains it better...

**Comment by indigoxela** [comment]
> Let's move our discussion here (from the PR)...

Good idea, this also came to my mind when seeing the PR comment thread growing. :+1: 

I didn't dig deeper yet, but I suspect, your proposal requires a bigger rewrite of function _system_rebuild_module_data(). And to my understanding this will cause a flood of log entries. For instance, after every cache flush or when enabling/disabling totally unrelated modules. If that's really the case, then I oppose. And that would for sure need another issue.

> In other words, let's either log these warning all the time, or not at all.

If these are the only options, then I'd personally prefer "not at all". But I don't think I'm the one to decide this.

One thing needs clarification: The skipping in the project installer only prevents the offending module to cause a PHP Notice, the module wouldn't be visible in that step anyway because of the missing type. Later on, in the full module list, it will be installable (my PR brings no change in this behavior).

If I get you right (not sure, though...) you'd prefer to keep that PHP notice?
The notice on its own doesn't provide much information (Undefined index type - huh?), it will just let Backdrop look broken.

**Comment by BWPanda** [comment]
I definitely don't prefer the PHP notice 🙂 

As for the rest, maybe others can weigh in. I'm not convinced of my argument anymore, but maybe some fresh eyes can take a look-see...

**Comment by klonos** [comment]
>> Let's move our discussion here (from the PR)...
>
>Good idea, this also came to my mind when seeing the PR comment thread growing. 👍

Yes, thanks 👍 

> I definitely don't prefer the PHP notice 🙂

Same here. I don't want PHP notices on the admin UI. Here's a few suggestions/thoughts:

1. I do find the reasoning behind moving the watchdog logging to `_system_rebuild_module_data()` legit. So long as these are errors logged in the dblog, it won't "make Backdrop look broken". My assumption is that these errors will eventually be discovered by more technical people (sys/site-admins instead of site builders), and then reported in the proper issue queue (we could make sure there's a link in the watchdog message).
2. If multiple similar errors are flooding the dblog, and we don't like that, we could then look into implementing a "this error was logged x more times" solution (separate issue though).
3. We should be "forcing" or "driving" contrib to explicitly specify the type of the project.
4. There are 2 possible scenarios for a submodule to NOT have its type defined:
   - It wasn't properly ported, in which case the maintainer should add `hidden` to it till they are ready to work on it.
   - The porting is complete, but setting the type was missed: Good! ...people will report it in the queue, and the maintainer (or the Bug Squad) will eventually fix it.
5. Instead of a watchdog message, we could be trowing a proper warning message, listing all the submodules/projects with a missing type. **It should be a single message**, like "the following projects do not specify their type:" ...followed by the list of projects in the same message (and perhaps individual links to report the issue).

**Comment by indigoxela** [comment]
It seems to me, we're stuck a bit here.

A suggestion: I revert the commit that prevents the "Undefined index type" notices. And we discuss the "type" problem in a follow-up issue. What do you think?

I wouldn't want to block the initial fix for the "Undefined index hidden" notices - which was an actual core bug. The "Undefined index type" thing is rather something that needs a little more time and discussion.

**Comment by BWPanda** [comment]
@indigoxela I think we're fine to have a PR here that *ignores* modules without a type, but leaves the *logging* of said modules for another issue. I'd hate to have a solution here that didn't end up hiding the front-end error message...

**Comment by indigoxela** [comment]
> I think we're fine to have a PR here that ignores modules without a type, but leaves the logging of said modules for another issue.

That seems like a good strategy. :+1: 

If that's also OK for @klonos, I'll remove the watchdog part from my PR. And @BWPanda could open the follow-up issue for the logging-when-what-where thing.

**Comment by indigoxela** [comment]
Done, the log message has been removed. My PR is ready for another review.

**Comment by BWPanda** [comment]
Code looks good. Didn't test it though.

**Comment by klonos** [comment]
Thanks @indigoxela ...I have both tested and reviewed this previously, and the only code change since has been the removal of the logging. So RTBC 👍

**Comment by quicksketch** [comment]
Great work folks! I like the new improved readability of the code and the fix. I merged https://github.com/backdrop/backdrop/pull/3330 into 1.x and 1.17.x.
